File: 02_insertAtMiddle/2.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newnode = (struct Node*)malloc(sizeof(struct Node));
    newnode->data = data;
    newnode->next = NULL;
    return newnode;
}

struct Node* insertInMiddle(struct Node* head, int x) {
    struct Node* newnode = createNode(x);

    if (head == NULL)
        return newnode;

    struct Node* slow = head;
    struct Node* fast = head;

    while (fast != NULL && fast->next != NULL) {
        fast = fast->next->next;
        if (fast == NULL)
            break;
        slow = slow->next;
    }

    newnode->next = slow->next;
    slow->next = newnode;

    return head;
}

void printList(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

struct Node* createLinkedList(int arr[], int n) {
    if (n == 0)
        return NULL;

    struct Node* head = createNode(arr[0]);
    struct Node* current = head;

    for (int i = 1; i < n; i++) {
        current->next = createNode(arr[i]);
        current = current->next;
    }

    return head;
}

int main() {
    int arr[] = {1, 2, 3, 4};
    int n = sizeof(arr) / sizeof(arr[0]);

    struct Node* head = createLinkedList(arr, n);

    printf("Original Linked List:\n");
    printList(head);

    int x = 99;
    head = insertInMiddle(head, x);

    printf("\nLinked List After Insertion:\n");
    printList(head);

    return 0;
}

File: 02_insertAtMiddle/2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def createLinkedList(arr):
    if not arr:
        return None
    head = Node(arr[0])
    current = head
    for val in arr[1:]:
        current.next = Node(val)
        current = current.next
    return head

def printLinkedList(head):
    current = head
    while current:
        print(current.data, end=" -> ")
        current = current.next
    print("None")

class Solution:
    def insertInMiddle(self, head, x):
        newnode = Node(x)
        
        if not head:
            return newnode
            
        slow = head
        fast = head
        while fast and fast.next:
            fast = fast.next.next
            if not fast: 
                break
            slow = slow.next
            
        newnode.next = slow.next
        slow.next = newnode
        
        return head

if __name__ == "__main__":
    arr = [1, 2, 3, 4]
    head = createLinkedList(arr)
    
    print("Original Linked List:")
    printLinkedList(head)

    x = 99
    sol = Solution()
    head = sol.insertInMiddle(head, x)
    
    print("\nLinked List After Insertion:")
    printLinkedList(head)